scipy/stats/_stats.pyx

```diff
@@ -753,3 +753,90 @@ def gaussian_kernel_estimate(points, values, xi, precision, dtype, real _=0):
                 estimate[j, k] += values_[i, k] * arg
 
     return np.asarray(estimate)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def gaussian_kernel_estimate_log(points, values, xi, precision, dtype, real _=0):
+    """
+    def gaussian_kernel_estimate_log(points, real[:, :] values, xi, precision)
+
+    Evaluate the log of the estimated pdf on a provided set of points.
+
+    Parameters
+    ----------
+    points : array_like with shape (n, d)
+        Data points to estimate from in ``d`` dimensions.
+    values : real[:, :] with shape (n, p)
+        Multivariate values associated with the data points.
+    xi : array_like with shape (m, d)
+        Coordinates to evaluate the estimate at in ``d`` dimensions.
+    precision : array_like with shape (d, d)
+        Precision matrix for the Gaussian kernel.
+
+    Returns
+    -------
+    estimate : double[:, :] with shape (m, p)
+        The log of the multivariate Gaussian kernel estimate evaluated at the
+        input coordinates.
+    """
+    cdef:
+        real[:, :] points_, xi_, values_, log_values_
+        real[:, :] whitening, estimate, estimate_, zeros_estimate_
+        real[:] max_, zeros_max
+        int i, j, k
+        int n, d, m, p
+        real arg, residual, log_arg, log_norm
+
+    n = points.shape[0]
+    d = points.shape[1]
+    m = xi.shape[0]
+    p = values.shape[1]
+
+    # Rescale the data
+    whitening = np.linalg.cholesky(precision).astype(dtype, copy=False)
+    points_ = np.dot(points, whitening).astype(dtype, copy=False)
+    xi_ = np.dot(xi, whitening).astype(dtype, copy=False)
+    values_ = values.astype(dtype, copy=False)
+
+    # Log transformation of values_. It is faster than np.log(values_).
+    log_values_ = np.empty((n, p), dtype)
+    for i in range(n):
+        for k in range(p):
+            log_values_[i, k] = math.log(values_[i, k])
+
+    # Evaluate the normalisation
+    log_norm = (- d / 2) * math.log(2 * PI)
+    for i in range(d):
+        log_norm += math.log(whitening[i, i])
+
+    # Create the result array and evaluate the weighted sum
+    estimate = np.zeros((m, p), dtype)
+    zeros_max = np.zeros((p,), dtype) - np.inf
+    empty_estimate_ = np.empty((p, n), dtype)
+    for j in range(m):
+        max_ = zeros_max.copy()
+        estimate_ = empty_estimate_.copy()
+        for i in range(n):
+            arg = 0
+            log_arg = 0
+            for k in range(d):
+                residual = (points_[i, k] - xi_[j, k])
+                arg += residual * residual
+
+            log_arg = -arg / 2 + log_norm
+            for k in range(p):
+                estimate_[k, i] = log_values_[i, k] + log_arg
+
+                # Collect max value for logsumexp
```

Comment:
I don't understand what is going on here to be honest - can you explain?

Reply:
It also has to do with `estimate`, `estimate_` and `empty_estimate_` - the copying and multiple arrays used there seem unnecessary. Why can this not be the same as for `gaussian_kernel_estimate`?

Reply:
> I don't understand what is going on here to be honest - can you explain?

When it returns
$$ \log(\sum^N_{i=1} \exp(x_i)) $$ 
I see an underflow error because the value of $ x $ is extremely small.
To avoid this problem the formula should be changed as 

$$ \begin{aligned}
\log(\sum^N_{i=1} \exp(x_i)) 
&= \log\{\exp(x_{max})\sum^N_{i=1} \exp(x_i - x_{max})\} \\
& = \log\{\sum^N_{i=1} \exp(x_i - x_{max})\}  + x_{max}
\end{aligned}
$$ 

By this conversion, the range of the target of $ \exp $ is $ [xmin−xmax, 0] $ so, the possibility of an underflow can be reduced.

This function exists as `logsumexp` in scipy. However, scipy's  `logsumexp` is implemented in Python, to keep the calculate fast, I wrote it by myself. In this code block you quoted, it tries to find $ x_{max} $ .


In the calculation of $ log $ world, multiplication changes to addition.
`gaussian_kernel_estimate`s'
```python
 for k in range(p):
    estimate[j, k] += values_[i, k] * arg
```
corresponds to 
`gaussian_kernel_estimate_log`s'
```python
for k in range(p):
    estimate_[k, i] = log_values_[i, k] + log_arg
```

___

> It also has to do with estimate, estimate_ and empty_estimate_ - the copying and multiple arrays used there seem unnecessary. Why can this not be the same as for gaussian_kernel_estimate?

For each `m`, I have to reset `estimate_`. If we didn't reset it, it would be possible that the previous iteration's max value would be selected.

I'm not sure my explanation has enough information. Does it make sense?

Reply:
Thanks @europeanplaice, yes that makes sense. It looks like this would have been much easier if we had `logsumexp` implemented in C/Cython and exposed in `special.cython_special`, right?

Reply:
Yes. It would be great not only for this PR, but also for future contributions!

Reply:
Hi, I am just passing by to look for any PR involving Cython in scipy.

We do have [a Cython implementation for `logsumexp` in scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/ef21ca1950f9570f2f6b1ce05d24111f78f9f612/sklearn/linear_model/_sag_fast.pyx.tp#L76-L95). I guess you can reuse it. 

On a side note, I am available for reviewing Cython PRs for scipy. So if needed, feel free to `@` me. :)

Reply:
Thank you @jjerphan that's good to know 😃 💪

Reply:
Do you mean a place where a person would propose some reviewing expertise? I would even wider this and say, someone proposes a certain expertise. Sometimes we have features on our wish list that we don't know much, or have deep technical questions that do not need software eng. expertise at all.

That would be awesome. We could do something similar to Publon and think about how to motivate people to engage and do this. This is definitely something that Scientific Python think about.

cc @jarrodmillman

Reply:
I've created [this thread on the Scientific Python Discourse](https://discuss.scientific-python.org/t/pooling-contributors-expertise-between-projects/381) to broaden the discussions regarding sharing skills across projects among the ecosystem.